Add cmake install target

markrtuttle opened the pull request:
Add an install target to the cmake build.

Aside from being generally useful, an install target for the purpose of simplifying maintenance was explicitly requested on the pull request to add CBMC to Homebrew (the MacOS package manager). https://github.com/Homebrew/homebrew-core/pull/53282

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ x] My PR is restricted to a single feature or bugfix.
- [ x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

Commit:
Add an install target to the cmake build.

Add an install target to the cmake build, and add a test of the
install target to travis.

hannes-steffenhagen-diffblue commented on src/goto-cc/CMakeLists.txt:
```diff
@@ -39,4 +40,17 @@ else()
       goto-cc $<TARGET_FILE_DIR:goto-cc>/goto-ld
     BYPRODUCTS ${CMAKE_BINARY_DIR}/bin/goto-ld)
+  install(TARGETS goto-cc DESTINATION bin)
+  INSTALL(CODE "execute_process( \
```
nitpick: This should be lowercase

markrtuttle replied:
Done.

hannes-steffenhagen-diffblue commented on src/goto-cc/CMakeLists.txt:
```diff
@@ -39,4 +40,17 @@ else()
+    ${CMAKE_INSTALL_PREFIX}/bin/goto-gcc \
+    )"
+  )
+  INSTALL(CODE "execute_process( \
```
this too

markrtuttle replied:
Done.

hannes-steffenhagen-diffblue commented on src/cbmc/CMakeLists.txt:
```diff
@@ -34,3 +34,10 @@ add_if_library(cbmc-lib jsil)
+
+# Man page
+if(NOT WIN32)
+  INSTALL(FILES ${CMAKE_SOURCE_DIR}/doc/man/cbmc.1
```
this too

markrtuttle replied:
Done.

hannes-steffenhagen-diffblue commented:
If you `include(GNUInstallDirs)` you can use `CMAKE_INSTALL_BINDIR`, `CMAKE_INSTALL_DOCDIR` etc. This is useful because for instance on linux you'd normally put docs not into `<install-root>/doc` but into `<install-root>/share/doc`, and those variables take care of those conventions.

markrtuttle commented:
@hannes-steffenhagen-diffblue I have implemented your recommendations.  Thanks.  I'm sorry for my delay, I had a deadline at work.

tautschnig reviewed (approved):
Looks good to me, with two wishlist-type suggestions:
1) I believe the three commits can now be squashed into a single commit.
2) Could we actually test this target in one of the Travis jobs?

markrtuttle commented:
Could someone show me the error log for cbmc-linux-cmake-gcc?  I'm trying to test the cmake install target.  I added "sudo cmake --install build" then "cbmc --version" to  buildspec-linux-cmake-gcc.yml.  Thanks.  @tautschnig @peterschrammel

chrisr-diffblue commented:
Looks like it failed with:
```
[Container] 2020/05/16 18:09:26 Running command sudo cmake --install build
/codebuild/output/tmp/script.sh: 4: /codebuild/output/tmp/script.sh: sudo: not found
```
I may be wrong :-) but I believe codebuild actually runs the whole build as 'root' - so I think you should be able to do the `cmake --install` without needing sudo.

markrtuttle commented:
@chrisr-diffblue I've removed sudo in the second commit.  Can you tell me the current cbmc-linux-cmake-gcc error?  Thanks.

chrisr-diffblue commented:
@markrtuttle Looks like this time it failed with:
```
[Container] 2020/05/19 21:23:36 Running command cmake --install build
CMake Error: The source directory "/codebuild/output/src176823731/src/github.com/diffblue/cbmc/build/build" does not exist.
Specify --help for usage, or press the help button on the CMake GUI.
```

hannes-steffenhagen-diffblue commented:
If you want to test the install functionality, you can also point CMake at an arbitrary directory by passing `-DCMAKE_INSTALL_PREFIX=<directory>` when first configuring, that way you can just install to a local directory and not have to worry about root rights or no root rights at all.

The `cmake --install build` line doesn't work because we're already `cd`ed into build, so it should just be `cmake --install .`

markrtuttle commented:
/bot run checks

markrtuttle commented:
I've modified the installation to use GNUInstallDirs.

I've extended the installation to create symbolic links to cbmc.1 for all binaries, as @tautschnig has done for the Debian packaging, until real man pages are written.

I've added a test of the installation (testing binaries and manpages) to Travis.

@tautschnig @chrisr-diffblue @hannes-steffenhagen-diffblue
